### PR TITLE
[App Service] `az webapp deploy`: Add `--track-status` arg to use deploymentstatus API

### PR DESCRIPTION
**Pull request by sarsharma**
**Related command**

`az webapp deploy --track-status`

**Description**

Adding support to allow use of deploymentstatus API to track the status for webapp deployments. The deploymentstatus API provides additional granular updates about the status of the app deployment. Keeping this additional verbose info behind an argument flag `--track-status`.
> Note: Initially, it will only be supported for Linux Webapps 

### Screenshots 
Site Started Successfully:
image

Build Failed: 
image

Site Failed to Start
image

Deploying to non-linux app 
image

Some instances failed to start
image

**Testing Guide**

Deploying webapps using CLI with command `az webapp deploy` along with arg `--track-status`

**History Notes**

[App Service] `az webapp deploy`: Add `--track-status` arg to use deploymentstatus API

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
App Service

File: src/azure-cli-core/azure/cli/core/util.py [hunk 1]
```diff
@@ -871,7 +871,7 @@ def check_connectivity(url='https://azure.microsoft.com', max_retries=5, timeout
 
 def send_raw_request(cli_ctx, method, url, headers=None, uri_parameters=None,  # pylint: disable=too-many-locals,too-many-branches,too-many-statements
                      body=None, skip_authorization_header=False, resource=None, output_file=None,
-                     generated_client_request_id_name='x-ms-client-request-id'):
+                     generated_client_request_id_name='x-ms-client-request-id', log_request_response=True):
```
Comment (vivekjilla) [inline src/azure-cli-core/azure/cli/core/util.py:874]: Do we need this `log_request_response` change checked in? If yes, we should keep the default value as `False`

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 2]
```diff
@@ -4176,6 +4176,46 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+        logger.info(deployment_status + "...")
+        if deployment_status == "RuntimeSuccessful":
+            runtime_successful = True
+            break
+        elif deployment_status == "RuntimeFailed":
+            raise CLIError("Deployment Failed: Runtime failed to start, check the deployment logs for more info")
+        elif deployment_status == "BuildFailed":
+            raise CLIError("Deployment Failed: Build failed, check the deployment logs for more info")
+    if not runtime_successful:
+        raise CLIError("Timeout reached while tracking deployment status, however, the deployment operation is still on-going. Navigate to your scm site to check the deployment status")
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4217]: I think we want to give more useful insights,in failure scenarios, like the logs endpoint, no. of instances updated, failed etc from the response, instead of throwing generic error. no?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 3]
```diff
@@ -4176,6 +4176,46 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+
+def _track_deployment_using_deploymentstatus_api(cmd, deploymentstatusapi_url, timeout=None):
+    total_trials = (int(timeout) // 2) if timeout else 450
+    num_trials = 0
+    runtime_successful = False
+    while num_trials < total_trials:
+        time.sleep(2)
+        response_body = send_raw_request(cmd.cli_ctx, "GET", deploymentstatusapi_url, log_request_response=False)
+        deployment_status = response_body.json().get('properties').get('status')
+        logger.info(deployment_status + "...")
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4208]: Would it be helpful to print time elapsed here, similar to how we do for SWA github action.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 4]
```diff
@@ -5113,6 +5156,22 @@ def _build_onedeploy_arm_url(params):
...
+    sub_id = get_subscription_id(params.cmd.cli_ctx)
+    if not params.slot:
+        base_url = (
+            f"subscriptions/{sub_id}/resourceGroups/{params.resource_group_name}/providers/Microsoft.Web/sites/"
+            f"{params.webapp_name}/deploymentStatus/{deployment_id}?api-version={client.DEFAULT_API_VERSION}"
+        )
+    else:
+        base_url = (
+            f"subscriptions/{sub_id}/resourceGroups/{params.resource_group_name}/providers/Microsoft.Web/sites/"
+            f"{params.webapp_name}/slots/{params.slot}/deploymentStatus/{deployment_id}"
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:5171]: nit: can we just add the `slot` details to the url, instead of constructing the entire url twice?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 5]
```diff
@@ -5208,8 +5267,15 @@ def _make_onedeploy_request(params):
...
         if poll_async_deployment_for_debugging:
             logger.info('Polling the status of async deployment')
+            if params.track_deployment.lower() == 'true':
+                # once deploymentstatus/latest is available, we can use it to track the deployment
+                deployment_id = _get_latest_deployment_id(params.cmd, params.resource_group_name, params.webapp_name, deployment_status_url, params.slot)
+                if deployment_id is None:
+                    raise CLIError("Unable to fetch deployment id for this deployment")
+                deploymentstatusapi_url = _build_deploymentstatus_url(params, deployment_id)
+                _track_deployment_using_deploymentstatus_api(params.cmd, deploymentstatusapi_url, params.timeout)
             response_body = _check_zip_deployment_status(params.cmd, params.resource_group_name, params.webapp_name,
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:5277]: we should probably not poll the old endpoint again, if we're tracking using the new one. No?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 6]
```diff
@@ -5208,8 +5267,15 @@ def _make_onedeploy_request(params):
         response_body = None
         if poll_async_deployment_for_debugging:
             logger.info('Polling the status of async deployment')
+            if params.track_deployment.lower() == 'true':
+                # once deploymentstatus/latest is available, we can use it to track the deployment
+                deployment_id = _get_latest_deployment_id(params.cmd, params.resource_group_name, params.webapp_name, deployment_status_url, params.slot)
+                if deployment_id is None:
+                    raise CLIError("Unable to fetch deployment id for this deployment")
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:5274]: if we can't find deployment_id, let's fallback to the old way instead of throwing error?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 7]
```diff
@@ -4176,6 +4176,46 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+    headers = get_scm_site_headers(cmd.cli_ctx, name, rg_name, slot)
+    total_trials = 30
+    num_trials = 0
+    while num_trials < total_trials:
+        time.sleep(2)
+        response = requests.get(deployment_status_url, headers=headers,
+                                verify=not should_disable_connection_verify())
+        try:
+            res_dict = response.json()
+        except json.decoder.JSONDecodeError:
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4192]: should we also catch any other exception too here and return None?

File: src/azure-cli-core/azure/cli/core/util.py (hunk 1 above)
Comment (sarsharma) [reply src/azure-cli-core/azure/cli/core/util.py]: By default it will print the entire request and response. Changing the parameter to `disable_request_response_logging` and setting default to `False`

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 5 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Yes, updating this

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 3 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Logging the additional info

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 2 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Logging the additional info

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 6 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:5295]: But the user explicitly wants to use deploymentstatus endpoint, And we have a reasonable number of retries to get the id, wouldn't it be better to fail instead here?

**Comment by yonzhan** [comment]
Please fix CI issues.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 6 above)
Comment (vivekjilla) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:5302]: yeah, makes sense. Then, maybe we can update the error message to also ask them to try without `track_deployment` flag?

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 8]
```diff
@@ -713,6 +713,7 @@ def load_arguments(self, _):
         c.argument('reset', help='Reset Java apps to the default parking page if set to true with no type specified.', arg_type=get_three_state_flag())
         c.argument('timeout', type=int, help='Timeout for the deployment operation in milliseconds. Ignored when using "--src-url" since synchronous deployments are not yet supported when using "--src-url"')
         c.argument('slot', help="The name of the slot. Default to the productions slot if not specified.")
+        c.argument('track_deployment', help="If true, web app deployment status will be tracked verbosely.", arg_type=get_three_state_flag())
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:716]: Maybe should we call this `track_runtime_status` or `poll_new_status` or something better?
let's get the review from @tulikac and @btardif  on the new param name and the general approach.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 9]
```diff
@@ -4179,6 +4179,73 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
     return res_dict
 
 
+def _get_latest_deployment_id(cmd, rg_name, name, deployment_status_url, slot):
+    import requests
+    from azure.cli.core.util import should_disable_connection_verify
+
+    headers = get_scm_site_headers(cmd.cli_ctx, name, rg_name, slot)
+    total_trials = 30
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4187]: 30 trials means, we wait for max one minute to get the deployment_id.. Isn't it too long wait time?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 10]
```diff
@@ -4179,6 +4179,73 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+    num_trials = 0
+    while num_trials < total_trials:
+        time.sleep(2)
+        response = requests.get(deployment_status_url, headers=headers,
+                                verify=not should_disable_connection_verify())
+        try:
+            res_dict = response.json()
+        except:   # pylint: disable=bare-except
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4195]: should we log the actual exception also in `debug` mode?

File: src/azure-cli-core/azure/cli/core/util.py [hunk 11]
```diff
@@ -865,7 +865,7 @@ def check_connectivity(url='https://azure.microsoft.com', max_retries=5, timeout
 
 def send_raw_request(cli_ctx, method, url, headers=None, uri_parameters=None,  # pylint: disable=too-many-locals,too-many-branches,too-many-statements
                      body=None, skip_authorization_header=False, resource=None, output_file=None,
-                     generated_client_request_id_name='x-ms-client-request-id'):
+                     generated_client_request_id_name='x-ms-client-request-id', disable_request_response_logging=False):
```
Comment (vivekjilla) [inline src/azure-cli-core/azure/cli/core/util.py:868]: check if we actually need this flag or how we're doing without it in other apis.

Comment (sarsharma) [reply src/azure-cli-core/azure/cli/core/util.py:868]: For use in other places, it is logging out the entire request & response header + content(for eg. if we do the deployment using `src-url`, it uses the ARM endpoint and uses this method to deploy. I think logging the headers and body will be too verbose for this use case, since we will be polling the deploymentstatus ARM endpoint every few seconds.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 9 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py:4187]: First kudu will return a temp deployment id, so it might be fine to wait for ~1min to get the actual deployment id. Also, this is just a temporary workaround to get the deployment id until we have `/deploymentstatus/latest` ARM API available.

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 12]
```diff
@@ -726,6 +727,7 @@ def load_arguments(self, _):
         c.argument('ignore_stack', options_list=['--ignore-stack'], help='If true, any stack-specific defaults are ignored.', choices=['true', 'false'])
         c.argument('timeout', options_list=['--timeout'], help='Timeout for the deployment operation in milliseconds.')
         c.argument('slot', help="The name of the slot. Default to the productions slot if not specified.")
+        c.argument('track_deployment', help="If true, web app deployment status will be tracked verbosely.", arg_type=get_three_state_flag())
```
Comment (ReaNAiveD) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:730]: I noticed that the PR also introduces the new argument for the command `az functionapp deploy`, but it is not mentioned in the PR description. Could you please update the PR description to include this information?

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py (hunk 8 above)
Comment (ReaNAiveD) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py:716]: Hi @sarsharma , is there any update on the param name? 

Please note that we are launching the release for this sprint this week. Please resolve all comments by 11/07/2023 10:00 UTC, otherwise the release of this PR will have to be postponed to the next sprint **(on 12-05)**.

Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py]: switching the param name to `track-runtime-status`

**Comment by zhoxing-ms** [comment]
@sarsharma Could you please resolve these CI issues?

**Comment by zhoxing-ms** [comment]
Could you please add some test cases for this PR?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 13]
```diff
@@ -4290,6 +4290,90 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+            return None
+        finally:
+            num_trials = num_trials + 1
+        if 'id' in res_dict and 'temp' not in res_dict['id']:
+            return res_dict['id']
+    return None
+
+
+def _track_deployment_runtime_status(params, deploymentstatusapi_url, timeout=None):
+    total_trials = (int(timeout) // 5) if timeout else 200
```
Comment (rupareddy5-21) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4318]: Did We check if timeout is working as expected?

Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: yeah, tested it and it works fine

File: src/azure-cli-core/azure/cli/core/util.py (hunk 11 above)
Comment (vivekjilla) [inline src/azure-cli-core/azure/cli/core/util.py:868]: let's undo the changes in this file

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 14]
```diff
@@ -4290,6 +4290,107 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+
+    if time_elapsed >= max_time_sec or deployment_status != "RuntimeSuccessful":
+        scm_url = _get_scm_url(params.cmd, params.resource_group_name, params.webapp_name, params.slot)
+        if deployment_status == "BuildInProgress":
+            deployments_log_url = scm_url + f"/api/deployments/{deployment_id}/log"
+            raise CLIError("Timeout reached while build was still in progress. "
+                           "Navigate to {} to check the build logs for your app.".format(
+                               deployments_log_url))
+        # For any other status, redirect user to /deployments/latest endpoint
+        latest_deployment_url = scm_url + "/api/deployments/latest"
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4382]: instead of latest, let's redirect them to the deployment_id

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 15]
```diff
@@ -5320,20 +5471,36 @@ def _make_onedeploy_request(params):
...
-            logger.info('Async deployment complete. Server response: %s', response_body)
+            logger.warning('Polling the status of %s deployment. Start Time: %s UTC',
+                           "async" if params.is_async_deployment else "sync",
+                           datetime.datetime.now(datetime.timezone.utc))
+            if params.track_runtime_status is not None and params.track_runtime_status:
+                # once deploymentstatus/latest is available, we can use it to track the deployment
+                deployment_id = _get_latest_deployment_id(params.cmd, params.resource_group_name,
+                                                          params.webapp_name, deployment_status_url, params.slot)
+                if deployment_id is None:
+                    raise CLIError("Unable to fetch deployment id for this deployment. "
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:5486]: instead of raising error, let's add a log saying couldn't track runtime status and fallback to kudu endpoint

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 16]
```diff
@@ -5348,8 +5515,11 @@ def _make_onedeploy_request(params):
 
     # check if an error occured during deployment
     if response.status_code:
-        raise CLIError("An error occured during deployment. Status Code: {}, Details: {}"
-                       .format(response.status_code, response.text))
+        scm_url = _get_scm_url(params.cmd, params.resource_group_name, params.webapp_name, params.slot)
+        latest_deployment_url = scm_url + "/api/deployments/latest"
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:5519]: latest --> deployment id

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py (hunk 12 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/_params.py]: Not including this flag for `functionapp deploy` now. Reverted changes

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py (hunk 16 above)
Comment (sarsharma) [reply src/azure-cli/azure/cli/command_modules/appservice/custom.py]: Here there is a possibility that we might not have a deploymentid since kudu build failed and deploymentid will be a temp id. Incase we don't have the deployment id, redirecting to /latest

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 17]
```diff
@@ -713,6 +713,7 @@ def load_arguments(self, _):
         c.argument('reset', help='Reset Java apps to the default parking page if set to true with no type specified.', arg_type=get_three_state_flag())
         c.argument('timeout', type=int, help='Timeout for the deployment operation in milliseconds. Ignored when using "--src-url" since synchronous deployments are not yet supported when using "--src-url"')
         c.argument('slot', help="The name of the slot. Default to the productions slot if not specified.")
+        c.argument('track_status', help="If true, web app runtime status during deployment will be tracked.", arg_type=get_three_state_flag())
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:716]: In the help, should we highlight, `deployment will be tracked for linux webapps.

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 18]
```diff
@@ -4296,6 +4297,117 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+            if int(deployment_properties.get('numberOfInstancesSuccessful')) > 0:
+                total_num_instances = int(deployment_properties.get('numberOfInstancesInProgress')) + \
+                    int(deployment_properties.get('numberOfInstancesSuccessful')) + \
+                    int(deployment_properties.get('numberOfInstancesFailed'))
+                error_text += "Site started with errors: {}/{} instances failed to start successfully\n".format(
+                    deployment_properties.get('numberOfInstancesFailed'),
+                    total_num_instances)
+            else:
+                error_text += "Deployment failed because the site failed to start within timeout limits.\n"
+                error_text += "InprogressInstances: {}, SuccessfulInstances: {}, FailedInstances: {}\n".format(
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4354]: Shall we print the instances details here only if total_instances > 0

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 19]
```diff
@@ -4296,6 +4297,117 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+            error_text = ""
+            if int(deployment_properties.get('numberOfInstancesSuccessful')) > 0:
+                total_num_instances = int(deployment_properties.get('numberOfInstancesInProgress')) + \
+                    int(deployment_properties.get('numberOfInstancesSuccessful')) + \
+                    int(deployment_properties.get('numberOfInstancesFailed'))
+                error_text += "Site started with errors: {}/{} instances failed to start successfully\n".format(
+                    deployment_properties.get('numberOfInstancesFailed'),
+                    total_num_instances)
+            else:
+                error_text += "Deployment failed because the site failed to start within timeout limits.\n"
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4353]: `start within timeout limits.` -> can we put the actual timeout limit used here?

File: src/azure-cli/azure/cli/command_modules/appservice/_params.py [hunk 20]
```diff
@@ -713,6 +713,7 @@ def load_arguments(self, _):
         c.argument('reset', help='Reset Java apps to the default parking page if set to true with no type specified.', arg_type=get_three_state_flag())
         c.argument('timeout', type=int, help='Timeout for the deployment operation in milliseconds. Ignored when using "--src-url" since synchronous deployments are not yet supported when using "--src-url"')
         c.argument('slot', help="The name of the slot. Default to the productions slot if not specified.")
+        c.argument('track_status', help="If true, web app runtime status during deployment will be tracked for linux web apps.", arg_type=get_three_state_flag())
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/_params.py:716]: nit: let's remove `runtime` and say `startup status`?

File: src/azure-cli/azure/cli/command_modules/appservice/custom.py [hunk 21]
```diff
@@ -4296,6 +4297,137 @@ def _check_zip_deployment_status(cmd, rg_name, name, deployment_status_url, slot
...
+    max_time_sec = int(timeout) if timeout else 1000
+    start_time = time.time()
+    time_elapsed = 0
+    deployment_status = None
+    response_body = None
+    while time_elapsed < max_time_sec:
+        try:
+            response_body = send_raw_request(params.cmd.cli_ctx, "GET", deploymentstatusapi_url).json()
+        except Exception as ex:  # pylint: disable=broad-except
+            logger.warning("Deployment status endpoint %s returned error: %s.", deploymentstatusapi_url, ex)
```
Comment (vivekjilla) [inline src/azure-cli/azure/cli/command_modules/appservice/custom.py:4340]: Should we only fallback incase of 404 not found exception? Also, emit this warning only incase of verbose maybe?

**Review by vivekjilla — APPROVED** [review]

**Review by tulikac — APPROVED** [review]
Status looks good

**Review by rupareddy5-21 — APPROVED** [review]

**Review by yanzhudd — APPROVED** [review]
